driver/reducecalculation_driver.hpp

```diff
@@ -128,17 +181,20 @@ class ReduceCalculationDriver : public Driver
 private:
     InputFlags inflags;
 
-    miopenTensorDescriptor_t inputDesc;
-    miopenTensorDescriptor_t outputDesc;
+    miopenTensorDescriptor_t inputDesc  = nullptr;
+    miopenTensorDescriptor_t outputDesc = nullptr;
 
-    std::unique_ptr<GPUMem> in_dev;
-    std::unique_ptr<GPUMem> out_dev;
-    std::unique_ptr<GPUMem> workspace_dev;
+    std::unique_ptr<GPUMem> in_dev        = nullptr;
+    std::unique_ptr<GPUMem> out_dev       = nullptr;
+    std::unique_ptr<GPUMem> workspace_dev = nullptr;
 
     std::vector<Tgpu> in;
     std::vector<Tgpu> out;
     std::vector<Tref> outhost;
 
+    std::vector<uint8_t> logical_out;
+    std::vector<uint8_t> logical_outhost;
```

Comment:
The reason for separating `out` and `logical_out` is to simplify handling logical calculations. If we merge out and `logical_out` into a single output, we would face much more complex logic and also need to introduce additional `base_arg`. To minimize changes to the driver, I opted to keep them separate.

Consider the scenario where out and `logical_out` are merged into a single output. The attributes of the `ReduceCalculationDriver` would look something like this:
```
std::vector<Tgpu> in;  
std::vector<Tgpu_out> out;  
std::vector<Tref> outhost; 
```

In numerical calculations, the `input` and `output` types are the same. However, for logical calculations, the output type is always `uint8_t`, which differs from the input type. This means we need to explicitly define `Tgpu_out`.

The real complication arises with `Tref`. For numerical calculations, `Tref` should be float or double to allow precision checks. For logical calculations, however, Tref should match the type of out (i.e., `uint8_t`). Managing this distinction within a unified structure would be cumbersome and error-prone, so separating them provides a cleaner and more manageable solution.

Reply:
I actually did do a version of merging those 2 outputs [[ref]](https://github.com/moreh-dev/MIOpen/pull/64/commits/f696296b397575f663efe901c7e537456ce3e89d). And it arises many problems, so I decided to keep them separately, to avoid changing too much on ReduceCalculationDriver APIs